### ISSUE TITLE
Incorrect case tree selection when comparing long Symbols during compiled Ruby code execution

### ISSUE THREAD

**Issue by jbaiza** [jruby/jruby#8421]
**Environment Information**

- JRuby versions 9.4.7.0 till 9.4.9.0 with empty JRUBY_OPTS
- MacOS Sequoia 15.0.1 with Apple M3 Max
- Java versions:
  - OpenJDK Runtime Environment (Zulu 8.74.0.17-CA-macos-aarch64) (build 1.8.0_392-b08)
  - OpenJDK Runtime Environment Temurin-11.0.21+9 (build 11.0.21+9)
  - OpenJDK Runtime Environment Zulu17.38+21-CA (build 17.0.5+8-LTS)

Other relevant info you may wish to add:

- Plain JRuby without any gems

**Expected Behavior**

Use the following files:

- test.rb:
```
class TestClass
  def select_case(a)
    case a
    when :zxcvbnmzxcvbnm
      1
    when :qwertyuiopqwertyuiop
      2
    when :asdfghjklasdfghjkl
      3
    when :a
      4
    when :z
      5
    end
  end
end

puts TestClass.new.select_case(:zxcvbnmzxcvbnm)
puts TestClass.new.select_case(:a)
puts TestClass.new.select_case(:z)
```

- test1.rb:
```
load 'test.class'
```

Compile the `test.rb` file using `jrubyc test.rb`.

Executing both files (`test.rb `and `test1.rb`), the same output should be displayed:
```
1
4
5
```

**Actual Behavior**

When the ruby file is executed using `jruby test.rb`, the expected lines are printed.
When the compiled class via the `test1.rb` file is executed using `jruby test1.rb`, the output is:
```
1
4

```

During our testing, we found that different symbol lengths gave different results. We faced this issue in our compiled app on customer environments where, in most cases, app restart fixes the issue.

A similar issue was previously #8157, but the sample test in it works with the latest versions.

If a quick fix isn't available for the issue, maybe there are recommendations on what we should avoid in case statements to bypass the problem.

**Comment by headius** [comment]
Ok so this is a problem loading pre-compiled Ruby code from `jrubyc`. The `jrubyc` command in JRuby 9.x does not actually compile to bytecode. Instead it compiles to our IR, serializes that into a byte array, and stores that in the .class. When loaded, it deserializes the IR and proceeds from there as though it were just parsed from Ruby.

I suspect this may be a bug in how it handles the optimized branching for a `case` with symbols. For some core types like numerics and Symbols, we will try to emit a fast jump table. I suspect that it is failing because it embeds the Symbol IDs, which are not guaranteed to be the same across runs.

I'm investigating that now.

**Comment by headius** [comment]
I have not confirmed the exact cause, since it *appears* that we do recalculate the Symbol IDs when deserializing the compiled IR, but I have a workaround for you.

The optimized switch-based `case/when` is only emitted for homogeneous cases, specifically all fixnum or all Symbol. Your example can be modified to not emit this optimization by adding a bogus non-symbol `when`:

```ruby
class TestClass
  def select_case(a)
    case a
    when :zxcvbnmzxcvbnm
      1
    when :qwertyuiopqwertyuiop
      2
    when :asdfghjklasdfghjkl
      3
    when :a
      4
    when :z
      5
    when "bogus"
    end
  end
end
```

This code will not compile with the optimized switch and should work when passed through jrubyc. Can't confirm at the moment, but give it a try.

**Comment by jbaiza** [comment]
Thank you for looking into this.
It seems that also `when nil` or `when false` will avoid optimization.
So, for now, we will go with this workaround.

**Comment by headius** [comment]
I believe I have found the issue.

The fix in #8157 only partially solved the issue. It rebuilds the jump table with new symbol IDs but it does not sort the table. If the new IDs are not sorted, then the binary search for the fast switch will fail to find some cases.

Enabling JVM assertions raises in the constructor for BSwitchInstr:

```
$ jruby -J-ea -e 'load "./blah.class"'
Unhandled Java exception: java.lang.AssertionError
java.lang.AssertionError: null
                              <init> at org/jruby/ir/instructions/BSwitchInstr.java:39
                              decode at org/jruby/ir/instructions/BSwitchInstr.java:121
```

Shouldn't be hard to fix.
